src/acp/rule.ts

```diff
@@ -197,3 +211,283 @@ export function setForbiddenRuleOnPolicy(
+ *
+ * @param url URL that identifies this [[Rule]].
+ */
+export function createRule(url: Url | UrlString): Rule {
+  const stringUrl = internal_toIriString(url);
+  let ruleThing = createThing({ url: stringUrl });
+  ruleThing = setUrl(ruleThing, rdf.type, acp.Rule);
+  return ruleThing;
+}
+
+/**
+ * ```{note} There is no Access Control Policies specification yet. As such, this
+ * function is still experimental and subject to change, even in a non-major release.
+ * ```
+ *
+ * Get the [[Rule]] with the given URL from an [[RuleDataset]].
+ *
+ * @param ruleResource The Resource that contains the given [[Rule]].
+ * @param url URL that identifies this [[Rule]].
+ * @returns The requested [[Rule]], if it exists, or `null` if it does not.
+ */
+export function getRule(
+  ruleResource: RuleDataset,
+  url: Url | UrlString
+): Policy | null {
+  const foundThing = getThing(ruleResource, url);
+  if (foundThing === null || getUrl(foundThing, rdf.type) !== acp.Rule) {
+    return null;
+  }
+  return foundThing;
+}
+
+/**
+ * ```{note} There is no Access Control Policies specification yet. As such, this
+ * function is still experimental and subject to change, even in a non-major release.
+ * ```
+ *
+ * List all the agents a [[Rule]] applies **directly** to. This will not include agents
+ * that are part of a group the [[Rule]] applies to, nor will it include specific agent
+ * classes, such as authenticated or public agents.
+ *
+ * @param rule The rule from which agents are read.
+ * @returns A list of the WebIDs of agents included in the rule.
+ * @since Unreleased
+ */
+export function getAgentForRuleAll(rule: Rule): WebId[] {
+  return getIriAll(rule, acp.agent).filter(
+    (agent: string) =>
+      agent !== acp.PublicAgent && agent !== acp.AuthenticatedAgent
+  );
+}
+
+/**
+ * ```{note} There is no Access Control Policies specification yet. As such, this
+ * function is still experimental and subject to change, even in a non-major release.
+ * ```
+ *
+ * Overrite the agents the [[Rule]] applies to with the provided agents.
+ *
+ * @param rule The rule for which agents are set.
+ * @param agents The list of agents the rule should apply to.
+ * @returns A copy of the input rule, applying to a different set of agents.
+ * @since Unreleased
+ */
+export function setAgentInRule(rule: Rule, agents: WebId[]): Rule {
+  // Preserve the special agent classes authenticated and public, which we
+  // don't want to overwrite with this function.
+  // NB: This raises the question wether the named agents and agent classes should
+  // be linked to with the same predicate.
+  const isPublic = hasPublicInRule(rule);
+  const isAuthenticated = hasAuthenticatedInRule(rule);
+  let result = cloneThing(rule);
+  if (agents.length === 0) {
+    result = removeAll(rule, acp.agent);
+  } else {
+    result = setIri(rule, acp.agent, agents[0]);
+    agents.slice(1).forEach((agent) => {
+      result = addIri(result, acp.agent, agent);
+    });
```

Comment:
I think this might be a bit easier to follow when not having to mentally split the agents array into two groups:

```suggestion
    result = removeAll(rule, acp.agent);
    agents..forEach((agent) => {
      result = addIri(result, acp.agent, agent);
    });
```

Reply:
Since we don't take an array of agents, it simplifies the function quite a bit.

Reply:
Fixed in e46e4f1f323ac62998356ad505726775a0b29467